{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3b29ca09",
   "metadata": {},
   "source": [
    "# Cell-Annotation\n",
    "\n",
    "###  Cell annotation by large language model GPT-4\n",
    "Hou, W., Ji, Z. Assessing GPT-4 for cell type annotation in single-cell RNA-seq analysis. Nat Methods 21, 1462–1465 (2024). https://doi.org/10.1038/s41592-024-02235-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e1ea3d48",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "from cell_annotator import CellAnnotator\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import os\n",
    "import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be1976b9",
   "metadata": {},
   "source": [
    "## Load Gene Orthology Reference\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b84b677d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NMR gene name</th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>Human gene stable ID</th>\n",
       "      <th>Human gene name</th>\n",
       "      <th>Mouse gene stable ID</th>\n",
       "      <th>Mouse gene name</th>\n",
       "      <th>Local_ID</th>\n",
       "      <th>Global_name</th>\n",
       "      <th>Global_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZMYND10</td>\n",
       "      <td>ENSHGLG00000010267</td>\n",
       "      <td>ENSG00000004838</td>\n",
       "      <td>ZMYND10</td>\n",
       "      <td>ENSMUSG00000010044</td>\n",
       "      <td>Zmynd10</td>\n",
       "      <td>LOCAL00000000000</td>\n",
       "      <td>ZMYND10</td>\n",
       "      <td>ENSHGLG00000010267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMIGO3</td>\n",
       "      <td>ENSHGLG00000010542</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOCAL00000000001</td>\n",
       "      <td>AMIGO3</td>\n",
       "      <td>ENSHGLG00000010542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TNFSF10</td>\n",
       "      <td>ENSHGLG00000010558</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOCAL00000000002</td>\n",
       "      <td>TNFSF10</td>\n",
       "      <td>ENSHGLG00000010558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSHGLG00000010617</td>\n",
       "      <td>ENSHGLG00000010617</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOCAL00000000003</td>\n",
       "      <td>ENSHGLG00000010617</td>\n",
       "      <td>ENSHGLG00000010617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NPRL2</td>\n",
       "      <td>ENSHGLG00000010655</td>\n",
       "      <td>ENSG00000114388</td>\n",
       "      <td>NPRL2</td>\n",
       "      <td>ENSMUSG00000010057</td>\n",
       "      <td>Nprl2</td>\n",
       "      <td>LOCAL00000000004</td>\n",
       "      <td>NPRL2</td>\n",
       "      <td>ENSHGLG00000010655</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        NMR gene name            gene_ids Human gene stable ID  \\\n",
       "0             ZMYND10  ENSHGLG00000010267      ENSG00000004838   \n",
       "1              AMIGO3  ENSHGLG00000010542                  NaN   \n",
       "2             TNFSF10  ENSHGLG00000010558                  NaN   \n",
       "3  ENSHGLG00000010617  ENSHGLG00000010617                  NaN   \n",
       "4               NPRL2  ENSHGLG00000010655      ENSG00000114388   \n",
       "\n",
       "  Human gene name Mouse gene stable ID Mouse gene name          Local_ID  \\\n",
       "0         ZMYND10   ENSMUSG00000010044         Zmynd10  LOCAL00000000000   \n",
       "1             NaN                  NaN             NaN  LOCAL00000000001   \n",
       "2             NaN                  NaN             NaN  LOCAL00000000002   \n",
       "3             NaN                  NaN             NaN  LOCAL00000000003   \n",
       "4           NPRL2   ENSMUSG00000010057           Nprl2  LOCAL00000000004   \n",
       "\n",
       "          Global_name           Global_ID  \n",
       "0             ZMYND10  ENSHGLG00000010267  \n",
       "1              AMIGO3  ENSHGLG00000010542  \n",
       "2             TNFSF10  ENSHGLG00000010558  \n",
       "3  ENSHGLG00000010617  ENSHGLG00000010617  \n",
       "4               NPRL2  ENSHGLG00000010655  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load gene otholog reference table (adapted with `gene_reference.ipynb`)\n",
    "gene_reference = pd.read_csv('/home/raquelcr/scanpy/gene_reference.csv')\n",
    "gene_reference.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f998b924",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>heterocephalus_glaber_female_gene</th>\n",
       "      <th>homo_sapiens_gene</th>\n",
       "      <th>orthology_type</th>\n",
       "      <th>perc_id</th>\n",
       "      <th>dn_ds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U6</td>\n",
       "      <td>ENSG00000252980</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>67.0000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3orf38</td>\n",
       "      <td>ENSG00000179021</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>85.1064</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZNF654</td>\n",
       "      <td>ENSG00000175105</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>90.5142</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGGBP1</td>\n",
       "      <td>ENSG00000163320</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>99.4012</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HTR1F</td>\n",
       "      <td>ENSG00000179097</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>93.9891</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12378</th>\n",
       "      <td>COX2</td>\n",
       "      <td>ENSG00000198712</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>69.1630</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12380</th>\n",
       "      <td>ND3</td>\n",
       "      <td>ENSG00000198840</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>67.8261</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12381</th>\n",
       "      <td>ND4</td>\n",
       "      <td>ENSG00000198886</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>69.4989</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12382</th>\n",
       "      <td>ND5</td>\n",
       "      <td>ENSG00000198786</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>65.1741</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12384</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>ortholog_one2one</td>\n",
       "      <td>76.3158</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11936 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      heterocephalus_glaber_female_gene homo_sapiens_gene    orthology_type  \\\n",
       "0                                    U6   ENSG00000252980  ortholog_one2one   \n",
       "1                               C3orf38   ENSG00000179021  ortholog_one2one   \n",
       "2                                ZNF654   ENSG00000175105  ortholog_one2one   \n",
       "3                                CGGBP1   ENSG00000163320  ortholog_one2one   \n",
       "4                                 HTR1F   ENSG00000179097  ortholog_one2one   \n",
       "...                                 ...               ...               ...   \n",
       "12378                              COX2   ENSG00000198712  ortholog_one2one   \n",
       "12380                               ND3   ENSG00000198840  ortholog_one2one   \n",
       "12381                               ND4   ENSG00000198886  ortholog_one2one   \n",
       "12382                               ND5   ENSG00000198786  ortholog_one2one   \n",
       "12384                              CYTB   ENSG00000198727  ortholog_one2one   \n",
       "\n",
       "       perc_id  dn_ds  \n",
       "0      67.0000    NaN  \n",
       "1      85.1064    NaN  \n",
       "2      90.5142    NaN  \n",
       "3      99.4012    NaN  \n",
       "4      93.9891    NaN  \n",
       "...        ...    ...  \n",
       "12378  69.1630    NaN  \n",
       "12380  67.8261    NaN  \n",
       "12381  69.4989    NaN  \n",
       "12382  65.1741    NaN  \n",
       "12384  76.3158    NaN  \n",
       "\n",
       "[11936 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nmr2human = pd.read_csv('/home/raquelcr/NMR-snRNA-seq/auto_annotation/nmr_to_human.csv')\n",
    "nmr2human = nmr2human[nmr2human['perc_id']>60] # filter out by perc_id\n",
    "nmr2human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "971582c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map mouse gene names to global names using gene_reference\n",
    "mouse_gene_map = dict(zip(gene_reference['Mouse gene name'], gene_reference['Global_name']))\n",
    "mouse_data.var['global_name'] = mouse_data.var_names.map(mouse_gene_map)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3077b09b",
   "metadata": {},
   "source": [
    "## Load NMR data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7a326675",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/raquelcr/scanpy/cellbender/denoised/NMR2_cerebral_cortex_denoised_filtered.h5',\n",
       " '/home/raquelcr/scanpy/cellbender/denoised/NMR1_cerebral_cortex_denoised_filtered.h5',\n",
       " '/home/raquelcr/scanpy/cellbender/denoised/NMR3_hippocampus_denoised_filtered.h5',\n",
       " '/home/raquelcr/scanpy/cellbender/denoised/NMR4_hippocampus_denoised_filtered.h5',\n",
       " '/home/raquelcr/scanpy/cellbender/denoised/NMR5_midbrain_denoised_filtered.h5',\n",
       " '/home/raquelcr/scanpy/cellbender/denoised/NMR6_midbrain_denoised_filtered.h5']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nmr_files = glob.glob('/home/raquelcr/scanpy/cellbender/denoised/*_filtered.h5')\n",
    "nmr_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b3bbe6b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading sample 2 of cerebral_cortex\n",
      "Loading sample 1 of cerebral_cortex\n",
      "Loading sample 3 of hippocampus\n",
      "Loading sample 4 of hippocampus\n",
      "Loading sample 5 of midbrain\n",
      "Loading sample 6 of midbrain\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[AnnData object with n_obs × n_vars = 19487 × 20774\n",
       "     obs: 'species', 'tissue', 'sample_id', 'replicate'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'species', 'global_name',\n",
       " AnnData object with n_obs × n_vars = 14268 × 20774\n",
       "     obs: 'species', 'tissue', 'sample_id', 'replicate'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'species', 'global_name',\n",
       " AnnData object with n_obs × n_vars = 6743 × 20774\n",
       "     obs: 'species', 'tissue', 'sample_id', 'replicate'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'species', 'global_name',\n",
       " AnnData object with n_obs × n_vars = 8179 × 20774\n",
       "     obs: 'species', 'tissue', 'sample_id', 'replicate'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'species', 'global_name',\n",
       " AnnData object with n_obs × n_vars = 7369 × 20774\n",
       "     obs: 'species', 'tissue', 'sample_id', 'replicate'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'species', 'global_name',\n",
       " AnnData object with n_obs × n_vars = 10080 × 20774\n",
       "     obs: 'species', 'tissue', 'sample_id', 'replicate'\n",
       "     var: 'gene_ids', 'feature_types', 'genome', 'species', 'global_name']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nmr_adatas = []\n",
    "nmr_gene_map = dict(zip(gene_reference['NMR gene name'], gene_reference['Global_name']))\n",
    "\n",
    "for file in nmr_files:\n",
    "    # CellBender output should be read with read_10x_h5\n",
    "    try: \n",
    "        adata = sc.read_10x_h5(file)\n",
    "    except:\n",
    "        adata = sc.read(file)\n",
    "\n",
    "    adata.var['species'] = 'nmr'\n",
    "\n",
    "    sample = file.replace('/home/raquelcr/scanpy/cellbender/denoised/NMR', '').replace('_denoised_filtered.h5', '').split('_', 1)\n",
    "    sample_id = sample[0]\n",
    "    tissue = sample[1]\n",
    "\n",
    "    print(f'Loading sample {sample_id} of {tissue}')\n",
    "\n",
    "    adata.obs['species'] = 'nmr'\n",
    "    adata.obs['tissue'] = tissue\n",
    "    adata.obs['sample_id'] = sample_id\n",
    "    adata.obs['replicate'] = (int(sample_id)+1)%2+1\n",
    "\n",
    "    # Map NMR gene IDs to global names\n",
    "    adata.var['global_name'] = adata.var_names.map(nmr_gene_map)\n",
    "    \n",
    "    nmr_adatas.append(adata)\n",
    "\n",
    "nmr_adatas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46ef7eac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change gene names to human gene names\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6225ddf",
   "metadata": {},
   "source": [
    "### Load few cells from human (adult and developing human) to perform cross-validation by triangulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1bd6666",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c140eb00",
   "metadata": {},
   "source": [
    "## GPT-Cell-Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a22943be",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_ann = CellAnnotator(\n",
    "    adata, species=\"human\", tissue=\"brain\", cluster_key=\"leiden\", sample_key=\"samples\",\n",
    ").annotate_clusters()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cellann",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
