{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "20c7bcbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NMR1_cerebral_cortex.h5ad      NMR4_hippocampus_raw.h5ad\n",
      "NMR1_cerebral_cortex_raw.h5ad  NMR5_midbrain_raw.h5ad\n",
      "NMR2_cerebral_cortex_raw.h5ad  NMR6_midbrain_raw.h5ad\n",
      "NMR3_hippocampus_raw.h5ad\n"
     ]
    }
   ],
   "source": [
    "!ls /home/raquelcr/scanpy/raw_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a1a0130",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pendingfiles=(\"NMR2_cerebral_cortex_raw.h5ad\" \"NMR3_hippocampus_raw.h5ad\" \"NMR4_hippocampus_raw.h5ad\" \"NMR5_midbrain_raw.h5ad\" \"NMR6_midbrain_raw.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dda66ad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p /home/raquelcr/scanpy/denoised"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e67af52",
   "metadata": {},
   "source": [
    "#### Work on tmux\n",
    "```bash\n",
    "tmux\n",
    "```\n",
    "\n",
    "#### Create cell bender environment (only if environment not already present)\n",
    "```bash\n",
    "raquelcr@aleph:~$ conda env create -f /home/raquelcr/seurat/cellbender/celbender-env.yml\n",
    "```\n",
    "\n",
    "#### Activate cellbender environment\n",
    "```bash\n",
    "conda activate cellbender-env\n",
    "```\n",
    "\n",
    "#### Make output directory\n",
    "```bash\n",
    "mkdir -p /home/raquelcr/scanpy/denoised\n",
    "```\n",
    "#### Run CellBender for every sample\n",
    "\n",
    "If you want only certain files use:\n",
    "```bash \n",
    "pendingfiles=(\"NMR2_cerebral_cortex_raw.h5ad\" \"NMR3_hippocampus_raw.h5ad\" \"NMR4_hippocampus_raw.h5ad\" \"NMR5_midbrain_raw.h5ad\" \"NMR6_midbrain_raw.h5ad\")\n",
    "```\n",
    "\n",
    "If you want all files use:\n",
    "```bash\n",
    "pendingfiles=ls /home/raquelcr/scanpy/raw_data\n",
    "```\n",
    "\n",
    "Then run:\n",
    "\n",
    "```bash\n",
    "for file in \"${pendingfiles[@]}\"; do\n",
    "    infile=\"/home/raquelcr/scanpy/raw_data/$file\"\n",
    "    outfile=\"/home/raquelcr/scanpy/denoised/${file/_raw.h5ad/_denoised.h5ad}\"\n",
    "    logfile=\"${outfile%.h5ad}.log\"\n",
    "\n",
    "    echo \"Processing $infile → $outfile\"\n",
    "    cellbender remove-background \\\n",
    "        --input \"$infile\" \\\n",
    "        --output \"$outfile\" \\\n",
    "        --total-droplets-included 20000 \\\n",
    "        > \"$logfile\" 2>&1\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00450ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc8f093f",
   "metadata": {},
   "source": [
    "#### Make output directory\n",
    "```bash\n",
    "mkdir -p /home/raquelcr/jcgorozco/denoised\n",
    "```\n",
    "#### Run CellBender for every sample\n",
    "\n",
    "If you want only certain files use:\n",
    "```bash \n",
    "pendingfiles=(\"NMR2_cerebral_cortex_raw.h5ad\" \"NMR3_hippocampus_raw.h5ad\" \"NMR4_hippocampus_raw.h5ad\" \"NMR5_midbrain_raw.h5ad\" \"NMR6_midbrain_raw.h5ad\")\n",
    "```\n",
    "\n",
    "If you want all files use:\n",
    "```bash\n",
    "pendingfiles=ls /home/raquelcr/jcgorozco/Axolotl_midbrain_sample1_STARsolo-out/GeneFull/raw\n",
    "```\n",
    "\n",
    "Then run:\n",
    "\n",
    "```bash\n",
    "for file in \"${pendingfiles[@]}\"; do\n",
    "    infile=\"/home/raquelcr/scanpy/raw_data/$file\"\n",
    "    outfile=\"/home/raquelcr/scanpy/denoised/${file/_raw.h5ad/_denoised.h5ad}\"\n",
    "    logfile=\"${outfile%.h5ad}.log\"\n",
    "\n",
    "    echo \"Processing $infile → $outfile\"\n",
    "    cellbender remove-background \\\n",
    "        --input \"$infile\" \\\n",
    "        --output \"$outfile\" \\\n",
    "        --total-droplets-included 20000 \\\n",
    "        > \"$logfile\" 2>&1\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85ba18ca",
   "metadata": {},
   "source": [
    "### Open filtered cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b9bc0bd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8553aa97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: /home/raquelcr/scanpy/clean_data/NMR1_cerebral_cortex.h_posterior.h5 Size (MB): 222.24975\n",
      "File: /home/raquelcr/scanpy/clean_data/NMR1_cerebral_cortex.h5 Size (MB): 39.354106\n",
      "File: /home/raquelcr/scanpy/clean_data/NMR1_cerebral_cortex_filtered.h5 Size (MB): 27.18458\n"
     ]
    }
   ],
   "source": [
    "files= glob.glob('/home/raquelcr/scanpy/clean_data/*h5')\n",
    "sizes_mb = [os.stat(file).st_size / 1_000_000 for file in files]   # MB (decimal)\n",
    "for i in range(len(files)):\n",
    "    print(f'File: {files[i]} Size (MB): {sizes_mb[i]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edaec440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3163481 × 20774\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample1_raw = sc.read_10x_h5('/home/raquelcr/scanpy/clean_data/NMR1_cerebral_cortex.h5')\n",
    "sample1_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5ec2f326",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 14268 × 20774\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample1_filtered = sc.read_10x_h5('/home/raquelcr/scanpy/clean_data/NMR1_cerebral_cortex_filtered.h5')\n",
    "sample1_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d648d317",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_ids         ENSHGLG00000004572\n",
       "feature_types       Gene Expression\n",
       "genome            HetGla_female_1.0\n",
       "Name: COX7C, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample1_filtered.var['genome']= 'HetGla_female_1.0'\n",
    "sample1_filtered.var.loc['COX7C'] # This gene was on the 'Table of top genes removed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "46afca8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample1_filtered.var_names_make_unique()\n",
    "sample1_filtered.obs['tissue']= 'cerebral_cortex'\n",
    "sample1_filtered.obs['replicate']= '1'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00011ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CGGGACTGTTGGCTAT-1', 'CCACGAGTCGCCGATG-1', 'TTTATGCCACCAACAT-1',\n",
       "       'ACCTGTCTCCGTGGCA-1', 'ATCGCCTAGCCTTCTC-1', 'CCTTTGGCACACACTA-1',\n",
       "       'CGTCCATAGAATAACC-1', 'AGGTCATTCAACACGT-1', 'CGAAGGAAGGATAATC-1',\n",
       "       'CTGGTCTTCAACTTTC-1',\n",
       "       ...\n",
       "       'GTCGTAAGTTAGAAAC-1', 'TACGGGCAGTACTCGT-1', 'TGAACGTGTGCAGGAT-1',\n",
       "       'TAATTCCAGTCGGGAT-1', 'TCATTCACATTGAAGA-1', 'TAAGCGTTCGTGGGTC-1',\n",
       "       'TGGGCGTAGGGCAGTT-1', 'ACGGAAGCACCGCTAG-1', 'GCTTTCGTCGGTCAGC-1',\n",
       "       'AACGGGAAGACCGCCT-1'],\n",
       "      dtype='object', length=14268)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample1_filtered.obs['n_counts'] = sample1_filtered.X.sum(axis=1).A1\n",
    "sample1_filtered.obs['n_genes'] = (sample1_filtered.X > 0).sum(axis=1).A1\n",
    "sample1_filtered.obs['percent_mito'] = np.sum(\n",
    "    sample1_filtered[:, sample1_filtered.var['gene_ids'].str.startswith('MT-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ac86f3e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00b4d1b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_list=[]\n",
    "for i in range(len(files)):\n",
    "    filtered_list.append(sc.read(files[i]))\n",
    "filtered_list[0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
